### ISSUE TITLE
Allow DependencyInjection in Normalizer-Methods

### ISSUE THREAD

**Issue by ChrisB9** [patchlevel/event-sourcing#361]
> This issue is meant more for discussion than for immediate action

If we have an event, that has e.g. a doctrine entity as a property, then it currently is not possible to get access to the data inside that event.

My aggregate only saves the ID to the payload. In my Normalizer I then retrieve that ID and turn that into an actual object. My thinking is, to use ProxyObjects that are going to be resolved, if the aggregate actually needs more than the ID.

Currently I am adding a static function to the entity that creates an entity with just its ID but from there I could actually never repopulate the entity data.

To support Injection, we'd just need a ContainerAware AttributeEventMetadataFactory, that, when calling normalizer() would then also add the necessary properties.

Is this Idea something useful for others?

**Comment by DanielBadura** [comment]
Could you perhaps provide an example Normalizer? I'm not sure if I understand the problem correctly. Also, could you provide how the event looks like?

**Comment by DanielBadura** [comment]
As of my understanding what you are doing is to retrieve the data for the event in the normalizer. This should not be done. The normalizer is only there to transform the event into a data format so it can be saved in for example a database.

You should put all the data you need in the event at the point you are creating and dispatching it. If you gather the data delayed changes already could have happened which then leads to an invalid event and thus to an invalid state.

We plan to update the normalizer api with https://github.com/patchlevel/event-sourcing/pull/346 for `3.x` of the library.

With these changes it would also be harder to allow DI in the normalizer as we think that there is no usecase for that. But if you can provide a valid usecase we will have a look into it.

**Comment by ChrisB9** [comment]
Sorry for replying late, and happy late new year

I am using normal entities and event sourced models side-by-side.
Lets say I have a `Member` Model as a normal entity and some `WorkingHistory` as an event model. Whenever the API saves something into the WorkingHistory, I do not want to save the entirety of User into the WorkingHistory-Table, that would completely bloat that table and the serialization would take forever.

I do agree, the normalizer might be actually the wrong place to populate the object as it would do that to every event and not just my most current one – but if the member data is only retrieved on access (proxied), then I should have no issue there and can fully interact with the result from my Aggregate or Projection.

This is how it currently looks like:

```php
public function __construct(
        #[Normalize(new WorkingHistoryJsonNormalizer())]
        public readonly WorkingHistoryEntry $entry,
        #[Normalize(new MemberNormalizer())]
        public readonly Member              $member,
    )
    {
    }
```

```php
public function denormalize(mixed $value): Member|null
    {
        if ($value === null) {
            return null;
        }
        if (is_string($value)) {
            // this just returns an empty Member Object with only a uuid
            return Member::fromUuid(Uuid::fromString($value));
        }
        throw new \Exception('some error');
    }
```

**Comment by DanielBadura** [comment]
Sorry for the late reply, was on vacation.

If I'm understanding everything correctly then you are doing here something really wrong. You have a doctrine entity `Member` which can be updated, these updates are tracked via `WorkingHistory` which is an `Aggregate` but itself has no functionality besides tracking events of a general event which is maybe called `MemberWasUpdated`. This is something we are not recommending to do, since `Member` and `WorkingHistory` are itself not atomic anymore and need each other to really have value.

In my opinion you should not do this kind of combination. In this case, I would recommend to either migrate `Member` to be a Aggregate with meaningful events where the eventstream is then the working history. Or the other way around and migrate the `WorkingHistory` to a logging concept.

Not allowing DI in the normalizers is also a guard to prevent such kind of misusage. If using the normalizers correctly only for transforming the event into a saveable format there is no need for extra services.

I hope i understood your usage correctly tho :D

**Comment by ChrisB9** [comment]
I think I finally understood why dependency injection is very wrong in the normalizers and now fully support making this very impossible.
Unless one uses the snapstore, each event will be deserialized every single time upon saving and if I do a database request every single time then the workload is intense.

My problem is probably that I designed the aggregate a bit too big and thus assuming I need to have all the information here. It is a very interesting learning curve coming from a standard ORM where I always have all information available. I have realized that every normalization comes with a cost, that with n-events hurts the performance. That was really good to understand and learn more about it and I thank you for your super valuable input - made me realize I do have to refactor in a second iteration :D
